modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessageFlusher.java

```diff
@@ -446,20 +444,63 @@ public int sendMessagesToSubscriptions(String destination, Set<AndesMessageMetad
                         //if we continue message order will break
                         break;
                     }
-                } else { //normal topic message
-                    if (numOfCurrentMsgDeliverySchedules == subscriptions4Queue.size()) {
+                } else {
+                    /**
+                     * For normal non-durable topic we pre evaluate room for all subscribers and if all subs has room
+                     * to accept messages we send them. This means we operate to the speed of slowest subscriber (to
+                     * prevent OOM). If it is too slow to make others fast, make that topic subscriber a durable
+                     * topic subscriber.
+                     */
+                    if(topicMessageDeliveryStrategy.equals(TopicMessageDeliveryStrategy.SLOWEST_SUB_RATE)) {
```

Comment:
can this be of use here? https://en.wikipedia.org/wiki/Strategy_pattern 
rather this runtime check? we can wire the strategy during the start up.